anvil/src/config.rs

```diff
@@ -818,131 +818,141 @@ impl NodeConfig {
                     )
                 };
 
-            let block = provider
-                .get_block(BlockNumber::Number(fork_block_number.into()))
-                .await
-                .expect("Failed to get fork block");
+                let block = provider
+                    .get_block(BlockNumber::Number(fork_block_number.into()))
+                    .await
+                    .expect("Failed to get fork block");
 
-            let block = if let Some(block) = block {
-                block
-            } else {
-                if let Ok(latest_block) = provider.get_block_number().await {
-                    panic!(
-                            "Failed to get block for block number: {fork_block_number}\nlatest block number: {latest_block}"
+                let block = if let Some(block) = block {
+                    block
+                } else {
+                    if let Ok(latest_block) = provider.get_block_number().await {
+                        let mut message = format!(
+                            "Failed to get block for block number: {fork_block_number}\n\
+                        latest block number: {latest_block}"
                         );
-                }
-                panic!("Failed to get block for block number: {fork_block_number}")
-            };
+                        // If the user is forking from an older block in a non-archive node, the
+                        // `get_block` call will return null.
+                        if fork_block_number < latest_block.as_u64() {
```

Comment:
Hmmm, I think you can fork from non-archive nodes as long as it's less than 64 or 128 blocks back—I don't quite remember. Have you tested this threshold?

Reply:
Yea it's 128 blocks with a regular full node, above that and you need an archive node

Reply:
Thanks for the input! I'm a little weary about checking for a difference of 128 blocks as it's an implementation detail that might change in `geth` and other clients may have different values. 

As I think more about, the check is actually superfluous, since at that point we know that the node couldn't return a block and the warning is phrased in a conditional and it advises only to take action if the issue persists. Wdyt about removing the check altogether as a resolution?

Reply:
sounds good!

Reply:
Hey thanks @Evalir! As I went ahead to make the change to remove the check, I realized that the check is indeed needed, but it should be a less than equal check. Let me expand why:

In this code we try to fetch a block in `anvil` to set up the fork. 

If the user specified a block number to fork from as a CLI argument, then we use that block number in the `eth_getBlockByNumber` call, otherwise we fetch the latest block number and use that in `eth_getBlockByNumber`.

Next we check if the `eth_getBlockByNumber` RPC call failed and if it did, we panic with the message `"Failed to get fork block"`.

Next, if the RPC call succeeded, we check if it returned `null` instead of the requested block and panic with the following message if it returned `null`:

```
Failed to get block for block number: N
latest block number: M
```

There are two reasons `eth_getBlockByNumber` can return `null` instead of the requested block:

1. The block number is greater than the latest block number.
1. The node is not an archive node and we're requesting an older block.

Now in this PR we want to log an additional warning if `eth_getBlockByNumber` returned `null` because the user is connecting to a non-archive node, but not if it returned `null`, because the user-supplied block number is greater than the latest block number.

We can achieve this by only logging the non-archive node warning if the requested block number is less than equal to the latest block number which is what I did in commit 07ebbc50d902516c397ed9b329151785dd1e4f66 now.

Note that if the user specifies a recent block number that even non-archive nodes can return, then `eth_getBlockByNumber` will not return `null`, so the non-archive node warning won't get logged.

Apologies for the confusion and hope this clears it up! Let me know if this resolution works for your please.

Reply:
Resolving per https://github.com/foundry-rs/foundry/pull/5131#pullrequestreview-1486460770